### PR TITLE
Memlet API revamp

### PR DESCRIPTION
**Pull request by tbennun**
- [x] Memlet API revamp
    * EmptyMemlet is removed in favor of Memlet() and `memlet.is_empty()`
    * Use legacy `Memlet.simple` API in all cases (Memlet API may be subject to further changes)
    * `num_accesses` is now two properties: `volume` and `dynamic` (for bounded but dynamic volume)
    * `wcr_conflict` is now `wcr_nonatomic` for clarity
    * New ease-of-use API for memlet construction from string (`Memlet('A[i:i+5]')`)
    * Memlet constructor accepts both strings and native types (AST, Subset) for supported fields
    * New `src_subset` and `dst_subset` properties
    * Remove unused `DYNAMIC` and `NA_DYNAMIC`

### CODE REVIEW

File: dace/memlet.py [hunk 1]
```diff
@@ -22,90 +24,196 @@ class Memlet(object):
...
+                 wcr_nonatomic: bool = False,
+                 allow_oob: bool = False):
+        """ 
+        Constructs a Memlet.
+        :param expr: A string expression of the this memlet, given as an ease
+                     of use API. Must follow one of the following forms:
+                     1. ``ARRAY``,
+                     2. ``ARRAY[SUBSET]``,
+                     3. ``ARRAY[SUBSET] -> OTHER_SUBSET``.
+        :param data: (DEPRECATED) Data descriptor name attached to this memlet.
```
Comment (definelicht) [inline dace/memlet.py:79]: :pagesofohdeer: Want to add a deprecated warning?

File: dace/memlet.py [hunk 2]
```diff
@@ -140,75 +248,178 @@ def simple(data,
...
-        return self.subset.bounding_box_size()
+        if self.src_subset:
+            return self.src_subset.bounding_box_size()
+        elif self.dst_subset:
+            return self.dst_subset.bounding_box_size()
+        return []
+
+    # New fields
+    @property
+    def src_subset(self):
```
Comment (definelicht) [inline dace/memlet.py:413]: Ohhh I see what you're doing there :P sneaking in the new definition

File: dace/memlet.py (hunk 1 above)
Comment (tbennun) [reply dace/memlet.py:79]: Only that parameter is deprecated. I don't know how to add a deprecated parameter.

File: dace/memlet.py (hunk 2 above)
Comment (tbennun) [reply dace/memlet.py:413]: yep :) slowly but surely

**Pull request merged**